### DISCUSSION TITLE
'horovodrun' gives '/bin/sh: 1: Syntax error: Unterminated quoted string'

### DISCUSSION THREAD

**Discussion by AllenHW** [horovod/horovod#3065] [Q&A]
I'm running `horovodrun` on a simple python script (for testing's sake). The environment is inside a custom docker container running on Ubuntu 16.04. Horovod version is _0.21.0_.

**`script.py`**:
```
print('testing')
```

\
Running:
```
horovodrun -np 2 python script.py
```
returns:
```
/bin/sh: 1: Syntax error: Unterminated quoted string
```

The script successfully runs with the equivalent openMPI command according to the [guide](https://horovod.readthedocs.io/en/stable/mpi_include.html):
```
mpirun -np 2 \
    -bind-to none -map-by slot \
    -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH \
    -mca pml ob1 -mca btl ^openib \
    python script.py
```

\
Any help would be appreciated!

**Comment by EnricoMi** [comment]
Can you run the following commands to get some more debug context:

    which horovod
    cat $(which horovod)
    which python

**Reply by AllenHW** [reply]
`which horovod` returns nothing. 

`which horovodrun` -> `/usr/local/bin/horovodrun`
`cat $(which horovodrun)` ->
```
#!/usr/bin/python3
# -*- coding: utf-8 -*-
import re
import sys
from horovod.runner.launch import run_commandline
if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(run_commandline())
```

`which python` -> `/usr/bin/python`
`which python3` -> `/usr/bin/python3`

Also:
`horovodrun -np 2` ->  `/bin/sh: 1: Syntax error: Unterminated quoted string`
`horovodrun -np 2 python3 script.py` -> `/bin/sh: 1: Syntax error: Unterminated quoted string`

**Reply by AllenHW** [reply]
`python3 --version` -> `Python 3.6.13`

**Reply by AllenHW** [reply]
I manually stepped through `/usr/local/bin/horovodrun` and it looks like `run_commandline()` is giving the error.

**Reply by EnricoMi** [reply]
can you add `--verbose` to the `horovodrun` command?

And can try all this with a released docker image, e.g. horovod/horovod:0.22.1?

**Reply by AllenHW** [reply]
Will run on the released `horovod/horovod:0.22.1` a bit later (will have to call it a day here) but an eager execution of the partial result :p is:

`horovodrun --verbose -np 2 python script.py` ->
```
Filtering local host names.
Remote host found: 
All hosts are local, finding the interfaces with address 127.0.0.1
Local interface found lo
mpirun --allow-run-as-root --tag-output -np 2 -H localhost:2 -bind-to none -map-by slot -mca pml ob1 -mca btl ^openib  -mca btl_tcp_if_include lo -x NCCL_SOCKET_IFNAME=lo  -x "TERM -x ARCHDIR -x AUTH0_CLIENT -x AWS_IAM_USER -x BAGFILE_SOURCE -x CMAKE_PREFIX_PATH -x CUDA_PKG_VERSION -x CUDA_VERSION -x CUDNN_VERSION -x DISABLE_GPU -x DQI_BASE_DIR -x EMBARK_DB -x EMBARK_ENV -x EUSDIR -x HOME -x HOROVOD_CCL_BGT_AFFINITY -x HOROVOD_GLOO_TIMEOUT_SECONDS -x HOROVOD_NUM_NCCL_STREAMS -x HOROVOD_STALL_CHECK_TIME_SECONDS -x HOROVOD_STALL_SHUTDOWN_TIME_SECONDS -x HOSTNAME -x JAVA_HOME -x LANG -x LD_LIBRARY_PATH -x LIBRARY_PATH -x LS_COLORS -x NCCL_VERSION -x NVIDIA_DRIVER_CAPABILITIES -x NVIDIA_REQUIRE_CUDA -x NVIDIA_VISIBLE_DEVICES -x OPENCV_THREAD_POOL_ACTIVE_WAIT_WORKER -x PATH -x PKG_CONFIG_PATH -x PS1 -x PWD -x PYTHONPATH -x ROSLISP_PACKAGE_DIRECTORIES -x ROS_DISTRO -x ROS_ETC_DIR -x ROS_MASTER_URI -x ROS_PACKAGE_PATH -x ROS_PYTHON_VERSION -x ROS_ROOT -x ROS_VERSION -x SHLVL -x TERM -x _  python scrit.p
```

The problem is in the corresponding `mpirun` command: `-x "TERM -x ARCHDIR` with a quote that doesn't get closed. Will follow up on what happens on the horovod container.

**Reply by EnricoMi** [reply]
Can you run `env` in your current setup? Maybe there is a environment variable calles `"TERM` which messes up the mpi command.

**Reply by AllenHW** [reply]
You were absolutely correct - there is:

`env` -> 
```
...
EMBARK_ENV=PRODUCTION
CUDA_VERSION=10.0.130
"TERM=xterm-256color"
SHLVL=1
HOME=/home/embark
...
```

Do you happen to know how to clear it? This isn't horovod related but
`unset \"TERM` actually returns ```bash: unset: `"TERM': not a valid identifier```

**Reply by EnricoMi** [reply]
Open a new terminal, if it is still there it must be in config file like `~/.bashrc` or `~/.profile`.
